### PR TITLE
0.7.0-dev (Go-based pump comms)

### PR DESCRIPTION
**Pull request by scottleibrand**
Prep for merge of 0.7.0-dev to dev. 

This is a fairly major change, which deprecates the Python-based `openaps` toolkit in favor of @ecc1's new Go-based pump communications libraries.  This dramatically improves the speed and reliability of pump communications, allowing OpenAPS ro run on a Pi Zero W (for example with the new Explorer HATs), but is also a big improvement on any other hardware as well.  Such a foundational change affects a lot of things, so we'll need some widespread testing of this branch by quite a few more folks, in addition to the testing that folks in the https://gitter.im/openaps/hardware-dev channel have been doing in the go-explorer-hat branch.  (Special thanks to @cluckj @alimhassam and @ecc1 for their extensive work on this, and to @zjedr @jpcunningh and @tmecklem for their contributions as well.)

Checkout and install the 0.7.0-dev branch per the directions at http://openaps.readthedocs.io/en/latest/docs/Customize-Iterate/update-your-rig.html#alternative-step-1b-test-a-feature-branch-not-recommended-for-initial-setup

Once you've confirmed everything is still working, we'll also need people to re-run oref0-setup, and even re-flash your rig and set up from scratch, to test all of the new Go-based pump comms stuff (replacing the python-based `openaps` toolkit).

As always, but **especially importantly for this update**, please make sure that you first test on a backup/dev rig, and that you keep running an unmodified 0.6.x rig while testing out all the new stuff until you're completely comfortable with it.

### CODE REVIEW

**Comment by alimhassam** [comment]
Medtronic cgm will need some testing here, and possibly some additional dev work, it is currently not supported by the go pump libraries.

**Comment by efidoman** [comment]
👍 Working well for me

**Comment by mgranberryto** [comment]
This isn't fetching a full 24h of pump history for my 715.  I suspect it might have been fixed upstream, but I'm in the process of getting a Go environment set up on my Edison to test.

**Comment by achris2000** [comment]
Brand new edison/explorer setup for testing, went through 0.6.0 master to 0.6.1-dev to 0.7.0-dev. The pump is a 522. Same settings as the primary 0.6.1-dev rig:
         "max_iob": 10,
        "max_daily_safety_multiplier": 4,
        "current_basal_safety_multiplier": 5,
        "autosens_max": 1.5,
        "autosens_min": 0.5,
        "A52_risk_enable": true,
        "rewind_resets_autosens": true,
        "adv_target_adjustments": true,
        "exercise_mode": true,
        "wide_bg_target_range": false,
        "sensitivity_raises_target": true,
        "high_temptarget_raises_sensitivity": true,
        "resistance_lowers_target": true,
        "low_temptarget_lowers_sensitivity": true,
        "unsuspend_if_no_temp": true,
        "enableSMB_always": true,
        "enableUAM": true,
        "curve": "rapid-acting",
        "maxSMBBasalMinutes": 30,
        "offline_hotspot": false
The 0.7.0-dev rig is looping (in parallel to the primary one) fine since yesterday without any errors sofar.

**Comment by mgranberryto** [comment]
I should have gotten back earlier, but I want to confirm that it is dangerous to loop without https://github.com/ecc1/medtronic/commit/8d1e447b7e35f00b9a26c4a3b5d01d56c13fefc6 present with an x15 pump.  Partial pump history, cut off at the date boundary, combined with any COB present at midnight, has lead to some inappropriately-delivered insulin.

**Comment by scottleibrand** [comment]
Do new installations get that automatically?  Or does something still need to be released?

**Comment by ecc1** [comment]
That fix is included in the v2018.04.27 release

**Comment by scottleibrand** [comment]
Has anyone does a fresh install since 2018.04.27 and confirmed that oref0-setup pulls in that release and fixes the midnight IOB issue?

**Comment by mgranberryto** [comment]
> Downloading Go pump binaries from: https://github.com/ecc1/medtronic/releases/download/v2018.04.27/medtronic-386-spi.tar.xz

It is pulling in a newer release which should fix the issue.

**Comment by cluckj** [comment]
I'm keeping a project log / feature request wiki here, please feel free to add to it if you have ideas!
https://github.com/openaps/oref0/wiki/Go-OpenAPS--Explorer-HAT-todo-list

**Comment by diabeticpilot** [comment]
I just updated to this branch and am encountering no issues after running the setup script. I'm using Logger to pull in BGs which is working well too.

**Comment by rollerdecom** [comment]
After switching from 0.6.1 and afterwards a fresh install after flashing, there seems both times to be no connection to the pump (WW-x12). In loop-log, cycling through "listening" and "Unsuccessful oref0-pump-loop".

**Comment by scottleibrand** [comment]
@rollerdecom you may be the first one to test 0.7.0 with an x12, so there may be legacy compatibility features missing.  Sounds like @cluckj would be interested in troubleshooting further in https://gitter.im/nightscout/intend-to-bolus

**Comment by rkresha** [comment]
missing a npm install -g along the way... after install seeing

`oref0-g4-loop: command not found`

to fix, 
```
cd ~/src/oref0 && npm install -g
```

**Comment by scottleibrand** [comment]
Was that from not following the directions carefully (and running `npm run global-install` after switching branches)? Or is there a reproducible problem with the install scripts or documentation?

**Comment by rkresha** [comment]
I have found the error in my ways, I wipe everything out when I re-run the setup, however I also run the quick-packages.sh script every time too which lays down 0.6.0 (or whichever) which then causes the script to not want to run `npm run global-install` in the if statement on line 569. 

Question, is there really any incentive to check if it is install already if you know you're not on master (line 572)? Why not just re-run `npm run global-install` it to make sure everything is gravy?  It's user error for sure, I just want to make sure I don't miss anything if `quick-packages.sh` changes. If you agree, I'll change it and submit a PR. 

Thoughts @scottleibrand ?

tl;dr: User error, but maybe room for improvement?

-Rob

**Comment by scottleibrand** [comment]
I believe that was a speed optimization (since `npm run global-install` takes awhile).  If you want to PR in a change to make it more robust, please do.

**Comment by scottleibrand** [comment]
I think this is about ready to merge to dev, as soon as we merge dev to master (https://github.com/openaps/oref0/issues/1007) and release that as 0.6.2.  Is there anything else that we'd like to explicitly test while this is still a feature branch before it gets merged for further testing in dev?

**Comment by tim2000s** [comment]
Installed this last night. Whilst it’s looping fine on the EXB/Edison, the onboard web page is not being populated with data. Haven’t had a chance to dig into why yet.

**Comment by jaylagorio** [comment]
I think we're a ways off from having this run on a Pi + HAT: https://github.com/openaps/oref0/issues/1035

**Comment by jaylagorio** [comment]
Having run setup a few times, most while the loop is running, I found that stopping cron before running setup significantly speeds things up. So: https://github.com/openaps/oref0/pull/1038

**Comment by zgohr** [comment]
Fresh install with raspbian lite bombs during setup with 

```
E: Unable to locate package tcpdump
E: Package 'screen' has no installation candidate
E: Unable to locate package acpid
E: Package 'vim' has no installation candidate
E: Package 'locate' has no installation candidate
E: Package 'ntpdate' has no installation candidate
```

I'm fairly certain it's because `apt-get update` didn't get run - there's a chance [this](https://github.com/openaps/oref0/blob/0.7.0-dev/bin/oref0-setup.sh#L715) `file_is_recent` conditional was truthy, meaning it doesn't run because it thinks it has recently. Unfortunately I can't be certain because neither of those echo statements got to my console.

The workaround is to `apt-get update` and re-run setup.

I know it's a pain to not fully understand something before fixing it - but I'm not sure why you wouldn't run `apt-get update` regardless of the last time it was run, either. That would be my suggestion here.

**Comment by jaylagorio** [comment]
Added the xdrip-js CGM system to installation: https://github.com/openaps/oref0/pull/1042

**Comment by scottleibrand** [comment]
@zgohr the recently-updated check was mostly an optimization for testing and debugging: it got annoying to wait for it to repeatedly re-do the same time-consuming thing that never accomplishes anything after the first time.  That said, most people won't be running oref0-setup repeatedly, so I'd be fine with merging a PR to comment out the recency check if folks think that'd be more reliable.

**Comment by tepidjuice** [comment]
> Installed this last night. Whilst it’s looping fine on the EXB/Edison, the onboard web page is not being populated with data. Haven’t had a chance to dig into why yet.

I'm having this issue as well. Haven't been able to figure out why.

**Comment by tim2000s** [comment]
That’s running okay for me now. Are you using the standard install directory?

**Comment by tepidjuice** [comment]
No I'm using a different install directory. However it should work with a non-standard name. It also doesn't work after changing the directory names in app.py.

**Comment by scottleibrand** [comment]
Since this is next-up for dev, and no one has reported any major blockers, I'd like to go ahead and merge this PR for further testing there, if no objections in the next day or so.
